File: src/my_ude_py/scripts/turn_camera_server.py
#!/usr/bin/env python3
import os
import cv2
import rclpy
from rclpy.node import Node
from ude_interface.srv import TurnCamera
from cv_bridge import CvBridge 
from std_msgs.msg import Float32

ANGLE_INITIAL_ = 0.0
FREQUENCY_DEFAULT_ = 1.0


class TurnCameraServerNode(Node): 
    def __init__(self):
        super().__init__("turn_camera_server") 

        self.declare_parameter("angle_initial", ANGLE_INITIAL_)
        self.declare_parameter("publish_frecuency", FREQUENCY_DEFAULT_)

        self.angle_= self.get_parameter("angle_initial").value
        self.publish_frecuency_=self.get_parameter("publish_frecuency").value

        self.available_angles = [-30, -15, 0, 15, 30]

        self.publisher_ = self.create_publisher(Float32, "angle", 10)
        self.timer = self.create_timer(1/self.publish_frecuency_, self.call_publish_angle)

        self.server_=self.create_service(TurnCamera, "turn_camera", self.callback_turn_camare)
        self.get_logger().info("Turn Camera Server Node has been started")

    def call_publish_angle(self):
         msg=Float32()
         msg.data= self.angle_
         self.publisher_.publish(msg)
    
    def callback_turn_camare(self, request, response):
        self.angle_=self.angle_-request.degree_turn
        if(self.angle_>30.0):
            self.angle_=30.0
        elif(self.angle_<-30.0):
            self.angle_=-30.0
        image = self.get_image(self.angle_)
        image_msg = CvBridge().cv2_to_imgmsg(image)
        response.camera_image = image_msg 
        return response 


    def get_image(self, angle):
        closest_angle = min(self.available_angles, key=lambda x:abs(x-angle))
        return self.read_in_image_by_filename(str(closest_angle) + ".png")

    def read_in_image_by_filename(self, file_name):
        dir_name = os.path.dirname(__file__)
        install_dir_index = dir_name.index("install/")
        file_location = dir_name[0:install_dir_index] + "src/my_ude_py/images/" + file_name 
        image = cv2.imread(file_location)
        return image 


def main(args=None):
    rclpy.init(args=args)
    node = TurnCameraServerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_ude_cpp/src/turn_camera_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include "ude_interface/srv/turn_camera.hpp"
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>
#include <cv_bridge/cv_bridge.h>

#include <chrono>

using namespace std::chrono_literals;

const float angle_initial_ = 0.0;
const double freq_default_ = 1.0;

using std::placeholders::_1;
using std::placeholders::_2;

class TurnCameraServerNode : public rclcpp::Node 
{
public:
    TurnCameraServerNode(std::string exe_dir) : Node("turn_camera_server")
    {
        ws_dir_ = get_ws_dir(exe_dir);
        this->declare_parameter("angle_initial",angle_initial_);
        angle_= this-> get_parameter("angle_initial").as_double();

        this-> declare_parameter("publish_frecuency", freq_default_);
        double publish_frecuency = this-> get_parameter("publish_frecuency").as_double();

        publisher_=this->create_publisher<std_msgs::msg::Float32>("angle",10);
        timer_=this->create_wall_timer(std::chrono::milliseconds((int) (1000/publish_frecuency)), 
                                        std::bind(&TurnCameraServerNode::callPublishAngle,this));
        
        server_ = this-> create_service<ude_interface::srv::TurnCamera>("turn_camera",
                std::bind(&TurnCameraServerNode::callbackTurnCamera, this, _1, _2));
    	RCLCPP_INFO(this->get_logger(), "Turn Camera Server Node has been started.");
    }

private:

    void callPublishAngle(){
        auto msg=std_msgs::msg::Float32();
        msg.data= angle_;
        publisher_->publish(msg);
    }

    void callbackTurnCamera(const ude_interface::srv::TurnCamera::Request::SharedPtr request,
                            const ude_interface::srv::TurnCamera::Response::SharedPtr response)
    {
        angle_=angle_-request->degree_turn;
        if(angle_<-30.0){
            angle_=-30.0;
        }else if (angle_>30.0)
        {
            angle_=30.0;
        }

        float closest_num = available_angles_[0];
		float smallest_angle = std::abs(angle_ - available_angles_[0]);
        float angle_diff;
        
        for (int i=1; i<5; i++) //Based on size of available_angles_ array
        {
            angle_diff = std::abs(angle_ - available_angles_[i]);
            if (angle_diff < smallest_angle)
            {
                smallest_angle = angle_diff;
                closest_num = available_angles_[i]; 
            }
        }
        std::string image_path = ws_dir_ + "src/my_ude_cpp/images/" + std::to_string((int)closest_num) + ".png";
        RCLCPP_INFO(this->get_logger(), "the path is %s", image_path.c_str());

        auto image = cv::imread(image_path);
        auto image_ptr = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", image).toImageMsg();

        response->camera_image = *image_ptr;
    }

    std::string get_ws_dir(std::string exe_dir)
        {
            std::string::size_type substr_index = exe_dir.find("install");
            return exe_dir.substr(0, substr_index);
        }
    float angle_;
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;

    rclcpp::Service<ude_interface::srv::TurnCamera>::SharedPtr server_;
    const float available_angles_ [5] {-30, -15, 0, 15, 30};
    std::string ws_dir_;

};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurnCameraServerNode>(argv[0]);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}